<a href="https://colab.research.google.com/github/AnneClaude/IDS_Project/blob/main/IDS_Analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### 1. Data Loading and Initial Setup
**Objective:** In this section, we will clone the repository files to access the author's built-in data download utilities. We will initialize the dataset loading pipeline for our chosen benchmark context to inspect its initial volume, columns, and layout.

In [1]:
# Clone the author's repository into our Google Colab environment
!git clone https://github.com/KostasEreksonas/IDS_test.git

# Move into the data directory to see what downloaders are available
%cd IDS_test/data
!ls -l

Cloning into 'IDS_test'...
remote: Enumerating objects: 972, done.
remote: Counting objects: 100% (97/97), done.
remote: Compressing objects: 100% (89/89), done.
remote: Total 972 (delta 19), reused 83 (delta 8), pack-reused 875 (from 1)
Receiving objects: 100% (972/972), 30.99 MiB | 11.75 MiB/s, done.
Resolving deltas: 100% (407/407), done.
/content/IDS_test/data
total 28
drwxr-xr-x 2 root root 4096 Jun 15 08:55 CIC_DDoS_2019
drwxr-xr-x 2 root root 4096 Jun 15 08:55 CIC_EV_2023
drwxr-xr-x 2 root root 4096 Jun 15 08:55 CIC_IDS_2017
drwxr-xr-x 2 root root 4096 Jun 15 08:55 CIC_UNSW_NB15
drwxr-xr-x 2 root root 4096 Jun 15 08:55 CSE_CIC_IDS_2018
drwxr-xr-x 2 root root 4096 Jun 15 08:55 LITNET_2020
drwxr-xr-x 3 root root 4096 Jun 15 08:55 NSL_KDD


### 2. Dataset Ingestion and Inspection (NSL-KDD)
**Objective:** We will execute the author's built-in data download script inside the `NSL_KDD` directory to pull the raw files. Once downloaded, we will load the dataset using Pandas to verify the data size, columns, and initial layout as required by the syllabus.

In [2]:
# Move into the NSL_KDD directory and look at its contents
%cd NSL_KDD
!ls -l

# Check if there is a script (usually named download.py or download.sh) and run it
# If the author has a python script, we can run it like this:
!python3 download.py

# Let's see what files appeared after running the downloader
!ls -l

/content/IDS_test/data/NSL_KDD
total 54612
-rw-r--r-- 1 root root      495 Jun 15 08:55 attack.types
drwxr-xr-x 2 root root     4096 Jun 15 08:55 csv
-rwxr-xr-x 1 root root       97 Jun 15 08:55 download_dataset.sh
-rw-r--r-- 1 root root    33503 Jun 15 08:55 index.html
-rw-r--r-- 1 root root      618 Jun 15 08:55 kddcup.names
-rw-r--r-- 1 root root     8648 Jun 15 08:55 KDDTest1.jpg
-rw-r--r-- 1 root root  1772643 Jun 15 08:55 KDDTest-21.arff
-rw-r--r-- 1 root root  1814092 Jun 15 08:55 KDDTest-21.txt
-rw-r--r-- 1 root root  3368089 Jun 15 08:55 KDDTest+.arff
-rw-r--r-- 1 root root  3441513 Jun 15 08:55 KDDTest+.txt
-rw-r--r-- 1 root root     8579 Jun 15 08:55 KDDTrain1.jpg
-rw-r--r-- 1 root root  3750763 Jun 15 08:55 KDDTrain+_20Percent.arff
-rw-r--r-- 1 root root  3822033 Jun 15 08:55 KDDTrain+_20Percent.txt
-rw-r--r-- 1 root root 18744510 Jun 15 08:55 KDDTrain+.arff
-rw-r--r-- 1 root root 19109424 Jun 15 08:55 KDDTrain+.txt
-rw-r--r-- 1 root root      115 Jun 15 08:55 README.md
pyt

#### Conclusions & Observations:
* **Redundant Download Step:** Running `download.py` throws a "No such file or directory" error because the author actually included the data files directly inside the repository under `NSL_KDD/`.
* **Data Availability:** The core benchmark datasets (`KDDTrain+.txt` and `KDDTest+.txt`) are fully present, meaning no external network fetching is required to continue.

### 3. Data Loading and Feature Inspection
**Objective:** We will load the main training file (`KDDTrain+.txt`) using Pandas. Since raw network logging dumps usually lack explicit column headers, we will inspect the first few rows to evaluate the schema layout, shape, and structure as required by the guidelines.

In [3]:
import pandas as pd

# Load the training file. It's comma-separated (.txt) and has no header row.
df = pd.read_csv('KDDTrain+.txt', header=None)

# Check the data size (Requirement 1.3)
print(f"Dataset Shape: {df.shape}\n")

# Inspect the first few rows (Requirement 1.1)
df.head()

Dataset Shape: (125973, 43)



,0,1,2,3,4,5,6,7,8,9,...,33,34,35,36,37,38,39,40,41,42
0,0,tcp,ftp_data,SF,491,0,0,0,0,0,...,0.17,0.03,0.17,0.00,0.00,0.00,0.05,0.00,normal,20
1,0,udp,other,SF,146,0,0,0,0,0,...,0.00,0.60,0.88,0.00,0.00,0.00,0.00,0.00,normal,15
2,0,tcp,private,S0,0,0,0,0,0,0,...,0.10,0.05,0.00,0.00,1.00,1.00,0.00,0.00,neptune,19
3,0,tcp,http,SF,232,8153,0,0,0,0,...,1.00,0.00,0.03,0.04,0.03,0.01,0.00,0.01,normal,21
4,0,tcp,http,SF,199,420,0,0,0,0,...,1.00,0.00,0.00,0.00,0.00,0.00,0.00,0.00,normal,21


#### Conclusions & Observations:
* **Dataset Scale:** The training subset contains 125,973 network connection records, each characterized by 43 attributes.
* **Lack of Features Schema:** The raw text lacks a header row, rendering the columns anonymous (0 to 42). To execute meaningful EDA and correlation matrices, we must map these to their domain-specific feature definitions.
* **Target Alignment:** Column 41 represents the classification target (e.g., `normal`, `neptune` attack types), while column 42 contains an auxiliary metric tracking instance classification difficulty, which may need to be dropped to prevent leakage.

### 4. Injecting Feature Names Schema
**Objective:** To perform compliant EDA and correlation measures, we will parse the feature definitions from the dataset metadata and apply them to our dataframe columns.

In [4]:
# The standard NSL-KDD column names list based on the kddcup.names file:
columns = [
    "duration", "protocol_type", "service", "flag", "src_bytes", "dst_bytes",
    "land", "wrong_fragment", "urgent", "hot", "num_failed_logins", "logged_in",
    "num_compromised", "root_shell", "su_attempted", "num_root", "num_file_creations",
    "num_shells", "num_access_files", "num_outbound_cmds", "is_host_login",
    "is_guest_login", "count", "srv_count", "serror_rate", "srv_serror_rate",
    "rerror_rate", "srv_rerror_rate", "same_srv_rate", "diff_srv_rate",
    "srv_diff_host_rate", "dst_host_count", "dst_host_srv_count",
    "dst_host_same_srv_rate", "dst_host_diff_srv_rate", "dst_host_same_src_port_rate",
    "dst_host_srv_diff_host_rate", "dst_host_serror_rate", "dst_host_srv_serror_rate",
    "dst_host_rerror_rate", "dst_host_srv_rerror_rate", "label", "difficulty_score"
]

# Assign columns to our dataframe
df.columns = columns

# Drop the auxiliary difficulty score column to avoid training bias
df = df.drop(columns=['difficulty_score'])

# Show the clean schema layout
print(f"Updated Dataset Shape: {df.shape}")
df[['protocol_type', 'service', 'flag', 'src_bytes', 'label']].head()

Updated Dataset Shape: (125973, 42)


,protocol_type,service,flag,src_bytes,label
0,tcp,ftp_data,SF,491,normal
1,udp,other,SF,146,normal
2,tcp,private,S0,0,neptune
3,tcp,http,SF,232,normal
4,tcp,http,SF,199,normal


#### Conclusions & Observations:
* **Successful Schema Mapping:** The table now properly reflects human-readable network attributes like `protocol_type`, `service`, `flag`, and `src_bytes`.
* **Proactive Feature Selection:** Dropped the `difficulty_score` feature to ensure our upcoming model testing avoids potential target leakages, guaranteeing evaluation integrity.

### 5. Exploratory Data Analysis (EDA) - Class Prevalence Analysis
**Objective:** We will calculate the empirical distribution of the target `label` column to determine the prevalence of normal traffic versus network attacks. We will analyze if the data suffers from heavy class skewness, which dictates our real-world choice of evaluation metrics (e.g., opting for F1/MCC over raw Accuracy).

In [5]:
# Calculate absolute counts and relative percentages for the top target labels
class_counts = df['label'].value_counts()
class_percentages = df['label'].value_counts(normalize=True) * 100

# Combine them into a clean overview dataframe
imbalance_report = pd.DataFrame({
    'Counts': class_counts,
    'Percentage (%)': class_percentages
})

print("--- Target Class Distribution Top 15 ---")
print(imbalance_report.head(15))

--- Target Class Distribution Top 15 ---
                 Counts  Percentage (%)
label                                  
normal            67343       53.458281
neptune           41214       32.716534
satan              3633        2.883951
ipsweep            3599        2.856961
portsweep          2931        2.326689
smurf              2646        2.100450
nmap               1493        1.185175
back                956        0.758893
teardrop            892        0.708088
warezclient         890        0.706501
pod                 201        0.159558
guess_passwd         53        0.042073
buffer_overflow      30        0.023815
warezmaster          20        0.015876
land                 18        0.014289


#### Conclusions & Observations (Class Imbalance Analysis):
* **Severe Long-Tailed Imbalance:** While benign traffic (`normal`) and DoS traffic (`neptune`) dominate more than 86% of the instances, critical system exploits like `buffer_overflow` (30 samples) and `land` attacks (18 samples) represent less than 0.03% of the data.
* **The Sampling Problem:** This represents a classic real-world cybersecurity data collection bias. In live network capture, background traffic or high-volume automated flood attacks naturally eclipse stealthy, targeted intrusions.
* **Evaluation Metric Constraints:** Because of this dramatic skew, using raw **Accuracy** as a performance benchmark is fundamentally flawed. A naive model that blindly predicts everything as `normal` or `neptune` can achieve ~86% accuracy while missing critical exploits. Thus, we must evaluate our upcoming models using robust class-agnostic metrics like **Matthews Correlation Coefficient (MCC)** and **F1-Score** to penalize failures on rare attack vectors.

### 6. Correlation Analysis & Methodological Justification
**Objective:** We will execute a pairwise correlation matrix over the numerical features to spot feature redundancies.
**Methodological Justification:** We select the **Spearman Rank Correlation** over the Pearson coefficient. Network traffic distributions (such as packet lengths or packet counts) typically adhere to non-linear, long-tailed power laws and contain prominent outliers. Pearson assumes normal distributions and strictly linear dynamics, whereas Spearman calculates monotonic relationships based on ranks, ensuring resilience against outliers and skewed non-normal continuous data structures.

In [6]:
# Select only numerical features for correlation calculation
numerical_df = df.select_dtypes(include=['float64', 'int64'])

# Calculate the Spearman correlation matrix
correlation_matrix = numerical_df.corr(method='spearman')

# Let's see which pairs have a near-perfect correlation (> 0.95), suggesting system redundancy
high_corr_pairs = correlation_matrix.unstack().sort_values(ascending=False)
high_corr_pairs = high_corr_pairs[high_corr_pairs < 1.0] # Remove self-correlations

print("--- Top 10 Redundant/Highly Correlated Numerical Feature Pairs (Spearman) ---")
print(high_corr_pairs.head(20))

--- Top 10 Redundant/Highly Correlated Numerical Feature Pairs (Spearman) ---
srv_serror_rate           serror_rate                 0.973119
serror_rate               srv_serror_rate             0.973119
rerror_rate               srv_rerror_rate             0.965777
srv_rerror_rate           rerror_rate                 0.965777
dst_host_srv_serror_rate  srv_serror_rate             0.942332
srv_serror_rate           dst_host_srv_serror_rate    0.942332
serror_rate               dst_host_serror_rate        0.935943
dst_host_serror_rate      serror_rate                 0.935943
dst_host_srv_serror_rate  serror_rate                 0.921663
serror_rate               dst_host_srv_serror_rate    0.921663
dst_host_srv_serror_rate  dst_host_serror_rate        0.919490
dst_host_serror_rate      dst_host_srv_serror_rate    0.919490
dst_host_same_srv_rate    dst_host_srv_count          0.919323
dst_host_srv_count        dst_host_same_srv_rate      0.919323
srv_serror_rate           dst_host_serro

#### Conclusions & Observations (Correlation & System Redundancy Analysis):
* **Identification of Severe System Redundancy:** We have mathematically spotted significant multi-collinearity. For example, `srv_serror_rate` and `serror_rate` share a correlation of ~0.973.
* **Cybersecurity Domain Context:** This redundancy occurs because of how network statistics are gathered. `serror_rate` tracks the percentage of connections that resulted in a SYN error across all traffic, while `srv_serror_rate` tracks the same error percentage but specifically scoped to the *same network service*. During a high-volume SYN flood (a Denial-of-Service attack like `neptune`), both metrics will skyrocket simultaneously, providing identical information to the model.
* **Tackling Redundancy:** Leaving these highly correlated features in the dataset causes feature bloat, increases model training times unnecessarily, and can degrade the interpretability of linear models. To tackle this in the feature engineering phase, we can drop one feature from each near-identical pair (e.g., dropping `srv_serror_rate` and keeping `serror_rate`) or use an automated dimensionality reduction technique like Principal Component Analysis (PCA).

### 7. Feature Engineering - Categorical Encoding
**Objective:** Machine learning algorithms operate strictly on mathematical vectors and cannot process string categories. We will encode our categorical metrics (`protocol_type`, `service`, `flag`) into numerical space.
**Methodological Justification:** We will use **One-Hot Encoding** for `protocol_type` and `flag` because they contain a limited number of distinct, unordered unique values. For `service`, which contains dozens of unique protocol strings, one-hot encoding would drastically inflate feature dimensionality; however, for tree-based models on this benchmark scale, it retains explicit structural clarity.

In [7]:
# Separate features (X) and string labels (y)
X = df.drop(columns=['label'])
y = df['label']

# Apply One-Hot Encoding to the categorical string columns
X_encoded = pd.get_dummies(X, columns=['protocol_type', 'service', 'flag'], drop_first=True)

# Convert all boolean flags generated by get_dummies to integers (0 or 1)
X_encoded = X_encoded.astype(float)

print(f"Original Feature Shape: {X.shape}")
print(f"One-Hot Encoded Feature Shape: {X_encoded.shape}")

Original Feature Shape: (125973, 41)
One-Hot Encoded Feature Shape: (125973, 119)


#### Conclusions & Observations (Categorical Encoding Analysis):
* **Dimensionality Expansion:** One-hot encoding converted our 3 string variables into 78 distinct binary indicator columns, pushing our feature matrix size from 41 to 119 variables.
* **Cybersecurity Interpretability Trade-off:** While expanding columns into sparse binary matrices enables numerical processing for linear algorithms, it fragments structural features (e.g., mapping separate protocols into disconnected indicators). However, for tree-based ensemble models, this preservation of categorical partitioning provides highly granular routing indicators for capturing subtle multi-stage attack profiles.

### 8. Model Training and Comparative Analysis
**Objective:** We will convert our multi-class target variable into a binary flag (`0` for `normal`, `1` for any malicious anomaly attack) to align with a standardized binary intrusion screening paradigm. We will split the dataset into training (80%) and validation (20%) sets using a fixed random seed. Then, we will train two separate algorithms—a Baseline Decision Tree and an Ensemble Random Forest—to contrast single-estimator boundaries against robust tree aggregation.

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, accuracy_score, matthews_corrcoef

# Step 1: Convert multi-class attack strings into binary labels (0 = normal, 1 = attack)
y_binary = y.apply(lambda label: 0 if label == 'normal' else 1)

# Step 2: Split data strictly into train (80%) and test (20%) with a set seed for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X_encoded, y_binary, test_size=0.2, random_state=42)

print(f"Training instances: {X_train.shape[0]} | Testing instances: {X_test.shape[0]}")

# Step 3: Train Model A - Decision Tree
dt_model = DecisionTreeClassifier(max_depth=10, random_state=42)
dt_model.fit(X_train, y_train)
dt_preds = dt_model.predict(X_test)

# Step 4: Train Model B - Random Forest (Using limited estimators for speed)
rf_model = RandomForestClassifier(n_estimators=50, max_depth=10, random_state=42, n_jobs=-1)
rf_model.fit(X_train, y_train)
rf_preds = rf_model.predict(X_test)

# Step 5: Comparative Evaluation Printout
print("\n================== DECISION TREE PERFORMANCE ==================")
print(f"Matthews Correlation Coefficient (MCC): {matthews_corrcoef(y_test, dt_preds):.4f}")
print(classification_report(y_test, dt_preds, target_names=['Normal', 'Anomaly']))

print("\n================== RANDOM FOREST PERFORMANCE ==================")
print(f"Matthews Correlation Coefficient (MCC): {matthews_corrcoef(y_test, rf_preds):.4f}")
print(classification_report(y_test, rf_preds, target_names=['Normal', 'Anomaly']))

Training instances: 100778 | Testing instances: 25195

================== DECISION TREE PERFORMANCE ==================
Matthews Correlation Coefficient (MCC): 0.9931
              precision    recall  f1-score   support

      Normal       1.00      1.00      1.00     13422
     Anomaly       1.00      1.00      1.00     11773

    accuracy                           1.00     25195
   macro avg       1.00      1.00      1.00     25195
weighted avg       1.00      1.00      1.00     25195


================== RANDOM FOREST PERFORMANCE ==================
Matthews Correlation Coefficient (MCC): 0.9926
              precision    recall  f1-score   support

      Normal       0.99      1.00      1.00     13422
     Anomaly       1.00      0.99      1.00     11773

    accuracy                           1.00     25195
   macro avg       1.00      1.00      1.00     25195
weighted avg       1.00      1.00      1.00     25195



#### Conclusions & Observations (Model Comparison):
* **Suspiciously High Performance:** Both the single Decision Tree (MCC: 0.9931) and the Ensemble Random Forest (MCC: 0.9926) achieved near-flawless precision and recall metrics.
* **The "Too Good to be True" Critique:** An MCC score this close to 1.0 suggests that the NSL-KDD tabular feature set contains deterministic indicators (such as particular combinations of `src_bytes` or `flag_S0`) that cleanly separate normal traffic from attack traffic in a closed dataset. In live network monitoring, shifting environments and novel zero-day attacks would heavily degrade this performance due to overfitting.

### 9. Error Analysis & Cybersecurity Trade-Offs
**Objective:** We will isolate the test instances where our Decision Tree model made an incorrect prediction. We will extract a confusion matrix and inspect the practical implications of False Positives versus False Negatives in a live Security Operations Center (SOC).

In [9]:
from sklearn.metrics import confusion_matrix

# Generate the Confusion Matrix
# Rows = Actual classes, Columns = Predicted classes
cm = confusion_matrix(y_test, dt_preds)
tn, fp, fn, tp = cm.ravel()

print("--- CONFUSION MATRIX ---")
print(f"True Negatives (Correctly caught Benign): {tn}")
print(f"False Positives (False Alarms): {fp}")
print(f"False Negatives (Missed Attacks!): {fn}")
print(f"True Positives (Correctly caught Attacks): {tp}\n")

# Find the indices of the rows where actual label doesn't match prediction
misclassified_indices = y_test[y_test != dt_preds].index

# Display a sample of the actual network packets our model failed on
print("--- SAMPLE OF MISCLASSIFIED PACKETS ---")
if len(misclassified_indices) > 0:
    error_sample = df.loc[misclassified_indices].copy()
    error_sample['predicted_label'] = [dt_preds[X_test.index.get_loc(idx)] for idx in misclassified_indices]
    print(error_sample[['protocol_type', 'service', 'flag', 'src_bytes', 'label', 'predicted_label']].head(10))
else:
    print("No errors found in this test slice.")

--- CONFUSION MATRIX ---
True Negatives (Correctly caught Benign): 13384
False Positives (False Alarms): 38
False Negatives (Missed Attacks!): 48
True Positives (Correctly caught Attacks): 11725

--- SAMPLE OF MISCLASSIFIED PACKETS ---
       protocol_type   service  flag  src_bytes            label  \
89442            udp   private    SF        100             nmap   
25228            tcp    finger    SF          6           normal   
10216            tcp  ftp_data    SF        334      warezclient   
27998            tcp    telnet  RSTR       9161           normal   
2968             udp   private    SF        100             nmap   
60775            tcp     other   REJ          0           normal   
121060           tcp     imap4    SF       1296           normal   
84226            tcp  ftp_data    SF          0  buffer_overflow   
72491            tcp    finger    S0          0           normal   
5579             tcp  ftp_data    SF          0  buffer_overflow   

        predict

#### Conclusions & Observations (Error & Trade-Off Analysis):
* **High-Risk Failures (False Negatives):** The model missed 48 active network attacks. Crucially, the error sample reveals that stealthy, low-volume exploits like `buffer_overflow` were classified as benign (`0`). In a practical application domain, a single missed buffer overflow can allow an adversary to execute arbitrary code, gain root access, and compromise entire subnets.
* **Operational Overhead (False Positives):** There were 38 false alarms where normal traffic (like standard `imap4` or `finger` queries) was flagged as malicious. In a real-world Security Operations Center (SOC), excessive false positives trigger alert fatigue, wasting expensive analyst hours investigating benign behavior.
* **The Asymmetric Security Trade-off:** This exposes why a simple tree structure optimized for overall accuracy is structurally fragile. In cybersecurity, the cost of a False Negative (a devastating breach) is orders of magnitude higher than a False Positive (a minor administrative nuisance). A robust security system must be tuned to minimize False Negatives, prioritizing high recall over a deceptive 99% baseline accuracy score.